crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -1023,18 +1031,104 @@ impl<'db> TypeInferenceBuilder<'db> {
         with_item: &ast::WithItem,
         definition: Definition<'db>,
     ) {
-        let expression = self.index.expression(&with_item.context_expr);
-        let result = infer_expression_types(self.db, expression);
-        self.extend(result);
+        let context_expr = self.index.expression(&with_item.context_expr);
+        self.extend(infer_expression_types(self.db, context_expr));
 
-        // TODO(dhruvmanila): The correct type inference here is the return type of the __enter__
-        // method of the context manager.
-        let context_expr_ty = self.expression_ty(&with_item.context_expr);
+        let target_ty = self.infer_context_expression(
+            &with_item.context_expr,
+            self.expression_ty(&with_item.context_expr),
+        );
 
         self.types
             .expressions
-            .insert(target.scoped_ast_id(self.db, self.scope()), context_expr_ty);
-        self.add_binding(target.into(), definition, context_expr_ty);
+            .insert(target.scoped_ast_id(self.db, self.scope()), target_ty);
+        self.add_binding(target.into(), definition, target_ty);
+    }
+
+    /// Infers the type of a context expression (`with expr`) and returns the target's type
+    ///
+    /// Returns [`Type::Unknown`] if the context expression doesn't implement the context manager protocol.
+    ///
+    /// ## Terminology
+    /// See [PEP343](https://peps.python.org/pep-0343/#standard-terminology).
```

Comment:
PEPs are historical documents that record the rationale and discussion around a change that was made to the language. They are not guaranteed to be up-to-date; it's better to link to the language specification where possible, which is living documentation that we work to keep up-to-date. Here I'd link to:

- https://docs.python.org/3/library/stdtypes.html#typecontextmanager
- https://docs.python.org/3/reference/datamodel.html#context-managers

(Don't ask me why this information is split over two separate pages!)